java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java

```diff
@@ -132,22 +132,56 @@ public BooleanConstraint shouldNotInverse() {
-    Constraint leftConstraint = initialProgramState.getConstraint(leftOp);
-    Constraint rightConstraint = initialProgramState.getConstraint(rightOp);
-    if (leftConstraint instanceof BooleanConstraint && rightConstraint == null && !shouldNotInverse().equals(booleanConstraint)) {
-      List<ProgramState> nullConstraints = copiedConstraints.stream().map(ps -> ps.addConstraint(rightOp, ObjectConstraint.nullConstraint())).collect(Collectors.toList());
+    BooleanConstraint leftConstraint = initialProgramState.getConstraint(leftOp, BooleanConstraint.class);
+    BooleanConstraint rightConstraint = initialProgramState.getConstraint(rightOp, BooleanConstraint.class);
+    if (leftConstraint != null && rightConstraint == null && !shouldNotInverse().equals(booleanConstraint)) {
+      List<ProgramState> nullConstraints = copiedConstraints.stream()
+        .flatMap(ps -> rightOp.setConstraint(ps, ObjectConstraint.NULL).stream())
+        .map(ps -> ps.addConstraints(rightOp, ps.getConstraints(rightOp).remove(BooleanConstraint.class))
+      ).collect(Collectors.toList());
       return ImmutableList.<ProgramState>builder().addAll(copiedConstraints).addAll(nullConstraints).build();
     }
     return copiedConstraints;
   }
 
-  @Override
   protected List<ProgramState> copyConstraint(SymbolicValue from, SymbolicValue to, ProgramState programState, BooleanConstraint booleanConstraint) {
     ProgramState newState = programState;
     if (programState.canReach(from) || programState.canReach(to)) {
       newState = programState.addConstraint(this, booleanConstraint);
     }
-    return super.copyConstraint(from, to, newState, booleanConstraint);
+    return copyConstraintFromTo(from, to, newState, booleanConstraint);
+  }
+
+  private List<ProgramState> copyConstraintFromTo(SymbolicValue from, SymbolicValue to, ProgramState programState, BooleanConstraint booleanConstraint) {
+    List<ProgramState> states = new ArrayList<>();
+    states.add(programState);
+    PMap<Class<? extends Constraint>, Constraint> leftConstraints = programState.getConstraints(from);
+    if (leftConstraints != null) {
+      leftConstraints.forEach((d, c) -> {
+        List<ProgramState> newStates = new ArrayList<>();
+        Constraint constraint = shouldNotInverse().equals(booleanConstraint) ? c : c.inverse();
+        states.forEach(state -> {
+          if (constraint == null) {
+            PMap<Class<? extends Constraint>, Constraint> constraints = state.getConstraints(to);
+            if (constraints != null) {
+              newStates.add(state.addConstraints(to, constraints.remove(c.getClass())));
+            } else {
+              newStates.add(state);
+            }
+          } else {
+            // special handling of copying inversed non-null constraint
+            if (ObjectConstraint.NULL == constraint && c != constraint) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1698)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1698)